### ISSUE TITLE
ChaCha20 cipher clears non-volatile xmm registers

### ISSUE THREAD

**Issue by Aleksandr-Melnikov** [openssl/openssl#12328]
If ChaCha20 cipher is called on processor with AVX-512 instruction support
https://github.com/openssl/openssl/blob/97ace46e11dba4c4c2b7cb67140b6ec152cfaaf4/crypto/chacha/asm/chacha-x86_64.pl#L2705
it clears content of the all xmm registers including non-volatile xmm6, xmm7, xmm8 during cleanup
https://github.com/openssl/openssl/blob/97ace46e11dba4c4c2b7cb67140b6ec152cfaaf4/crypto/chacha/asm/chacha-x86_64.pl#L2836

Issue is reproduced with openssl 1.1.1c, but it seems it should be the same on master.

To reproduce on processor supporting AVX-512VL:

```
...
asm ("movaps %xmm1, %xmm8"); //to ensure that xmm8 is not zero
SSL_CTX_set_cipher_list(ssl_ctx, "ECDHE-RSA-CHACHA20-POLY1305"); // ensure using ChaCha20
...
/*
Perform ssl request to www.freebsd.org and check that xmm8 is now zero
*/
```

**Comment by davidben** [comment]
Looks like the issue is that this instruction touches register 8:
https://github.com/openssl/openssl/blob/97ace46e11dba4c4c2b7cb67140b6ec152cfaaf4/crypto/chacha/asm/chacha-x86_64.pl#L2495

But the Windows-specific register save and restore logic only handles xmm6 and xmm7.
https://github.com/openssl/openssl/blob/97ace46e11dba4c4c2b7cb67140b6ec152cfaaf4/crypto/chacha/asm/chacha-x86_64.pl#L2517-L2520

I'm guessing this slipped through because the `vprold` call doesn't use `$a`, `$b`, etc.

**Comment by Aleksandr-Melnikov** [comment]
Thanks, @davidben, I believe you are right.

**Comment by t8m** [comment]
This was fixed by #15315 before the 3.0 release.
